clusterloader2/pkg/provider/provider.go

```
@@ -0,0 +1,96 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package provider
+
+import (
+	"fmt"
+	"strings"
+)
+
+type InitOptions struct {
+	KubemarkRootKubeConfigPath string
+	ProviderName               string
+	ProviderConfigs            []string
+}
+
+type Provider interface {
+	Name() string
+	CanEnablePrometheusServer() error
+	// SSHToMasterSupported determines whether SSH access to master machines is possible.
+	// If false (impossible for many  providers), ClusterLoader will skip operations requiring it.
+	// Default value is false.
+	SSHToMasterSupported() bool
+	// Default value is true.
+	CanAccessAPIServerPprofEndpoint() bool
+	GetPortForComponent(componentName string) (int, error)
+	GetProtocolForComponent(componentName string) string
+	RootFrameworkKubeConfigOverride() string
+	RunSSHCommand(cmd, host string) (stdout string, stderr string, exit int, err error)
+
+	// If snapshot prometheus disk is supported.
+	// Default is false.
+	CanSnapshotPrometheusDisk() bool
+	// Default is false.
+	PrometheusScrapeApiserverOnly() bool
+	// Default is false.
+	SupportNodeKiller() bool
+	// Default is true.
+	CanGrabMetricsFromKubelets() bool
+}
+
+func NewProvider(initOptions *InitOptions) Provider {
+	configs := parseConfigs(initOptions.ProviderConfigs)
+	if initOptions.KubemarkRootKubeConfigPath != "" {
+		configs[RootKubeConfigKey] = initOptions.KubemarkRootKubeConfigPath
+	}
+	switch initOptions.ProviderName {
+	case "aks":
+		return NewAKSProvider(configs)
+	case "aws":
+		return NewAWSProvider(configs)
+	case "eks":
+		return NewEKSProvider(configs)
+	case "gce":
+		return NewGCEProvider(configs)
+	case "gke":
+		return NewGKEProvider(configs)
+	case "kubemark":
+		return NewKubemarkProvider(configs)
+	case "local":
+		return NewLocalProvider(configs)
+	case "skeleton":
+		return NewSkeletonProvider(configs)
+	case "vsphere":
+		return NewVsphereProvider(configs)
+	default:
+		return &DefaultProvider{}
```

Let's return an error here (i.e. make it obligatory to explicitly specify provider), otherwise it may lead to situation when I forgot to define provider and I run into unexpected issues in my tests. E.g. I run 12h load test on 5k load cluster and because default provider was used the prometheus monitoring was disabled and I realize that after the test was run.

---

It's only there because currently provider is not a required field and I was not sure about changing that. Sounds reasonable to make it required.